### ISSUE TITLE
Switching a large blob store to Storj is problematic

### ISSUE THREAD

**Issue by rjkroege** [kopia/kopia#3426]
I attempted to switch to Storj ([Globally Distributed Cloud Object Storage](https://www.storj.io/)) as a blobstore. This doesn't currently work very well. Two observations about this:

* `kopia repository sync-to  rclone` seems to work but is very slow
* the Storj S3 compatibility interface works just fine for *small* blobstores but is not adequate for production use because the Storj S3 gateway limits `s3_storage.go:/ListBlobs()` to a thousand blobs (per documentation) total if using prefix matching.

I see two solutions:

* run a local Storj S3 gateway with much larger limits
* write a "native" Storj blob implementation in Kopia.

I would expect that the "native" implementation (i.e. Storj blob implementation) is preferable for its performance and robustness. 

@jkowalski et al: are you open to PRs implementing such a feature? If so, I will attempt to write this.

**Comment by mathrick** [comment]
Any progress? This makes backups on Storj work fine at first, but then stop working as time goes on, which is very, very bad. I just discovered that my backups have been silently failing for a while now, and I can't connect to the repository anymore, with errors during connection initialisation:

```
20:06:03.690 finished initial cache scan {"cache":"contents","duration":"3.34742068s","totalRetainedSize":5242839371,"tooRecentBytes":0,"tooRecentCount":0,"maxSizeBytes":5242880000,"limitBytes":0,"inUsePercent":99}
20:06:03.965 finished initial cache scan {"cache":"metadata","duration":"268.039301ms","totalRetainedSize":5241761666,"tooRecentBytes":0,"tooRecentCount":0,"maxSizeBytes":5242880000,"limitBytes":0,"inUsePercent":99}
20:06:06.857 finished initial cache scan {"cache":"index-blobs","duration":"2.891745439s","totalRetainedSize":4094244378,"tooRecentBytes":0,"tooRecentCount":0,"maxSizeBytes":5242880000,"limitBytes":0,"inUsePercent":78}
20:06:06.869 [STORAGE] concurrency level reached {"maxConcurrency":1}
20:06:06.887 [STORAGE] concurrency level reached {"maxConcurrency":2}
20:06:06.891 [STORAGE] concurrency level reached {"maxConcurrency":3}
20:06:06.897 [STORAGE] concurrency level reached {"maxConcurrency":4}
20:06:12.604 [STORAGE] ListBlobs {"prefix":"xs","resultCount":0,"error":"Please reduce your request rate.","duration":"5.707400709s"}
20:06:12.606 [STORAGE] ListBlobs {"prefix":"xw","resultCount":0,"error":"context canceled","duration":"5.722143519s"}
20:06:12.606 [STORAGE] ListBlobs {"prefix":"xe","resultCount":0,"error":"context canceled","duration":"5.718712631s"}
20:06:12.606 [STORAGE] ListBlobs {"prefix":"xr","resultCount":0,"error":"context canceled","duration":"5.714953374s"}
20:06:19.087 [STORAGE] ListBlobs {"prefix":"xw","resultCount":0,"error":"Please reduce your request rate.","duration":"6.378387646s"}
20:06:19.088 [STORAGE] ListBlobs {"prefix":"xs","resultCount":0,"error":"context canceled","duration":"6.374979881s"}
20:06:19.088 [STORAGE] ListBlobs {"prefix":"xr","resultCount":0,"error":"context canceled","duration":"6.380425271s"}
20:06:19.088 [STORAGE] ListBlobs {"prefix":"xe","resultCount":0,"error":"context canceled","duration":"6.379743816s"}
20:06:25.028 [STORAGE] ListBlobs {"prefix":"xs","resultCount":0,"error":"Please reduce your request rate.","duration":"5.784996289s"}
20:06:25.029 [STORAGE] ListBlobs {"prefix":"xe","resultCount":0,"error":"context canceled","duration":"5.788615194s"}
20:06:25.029 [STORAGE] ListBlobs {"prefix":"xr","resultCount":0,"error":"context canceled","duration":"5.78865443s"}
20:06:25.033 [STORAGE] ListBlobs {"prefix":"xw","resultCount":0,"error":"context canceled","duration":"5.792762418s"}
[...]
```

**Comment by rjkroege** [comment]
I am so sorry that this has not been working fo you. That's indeed very bad. It's not landed yet because as @jkowalski pointed out in the review, it's not done. 

I have been using this for all of my backups for a while and on small memory machines I too see "work fine at first, but then stop working as time goes on". However, the files successfully saved (at least for me) all seem (albeit sometimes slowly) retrievable. I am doing something with releasing storj resources. I have not yet fixed. I plan to get to this done but seem to keep on having other stuff to do. Feel free to pester me intermittently and I will try harder to make this a higher priority.

**Comment by mathrick** [comment]
@rjkroege: thank you for the update. I do sympathise with your situation, I also seem to always have 10x important fixes to make than the time to make them. That said, a fix would be incredibly helpful, so I will definitely pester you in the future :)

**Comment by ptandler** [comment]
After using kopia & storj for several years now successfully, I recently also hit the "ListObjects(V2): listing too many items for gateway-side filtering using arbitrary delimiter/prefix" limit. Luckly, I do a second backup to HD.

@rjkroege how likely do you intend to continue with the native storj implementation? What about asking a friendly AI to generate a draft for the tests?

**Comment by ptandler** [comment]
In the [StorJ forum](https://forum.storj.io/t/error-listing-objects-listing-too-many-items-for-gateway-side-filtering/30296/6) they suggest to try rclone - or a different backup tool 😬

**Comment by rjkroege** [comment]
I intend to finish this but it's unfortunately behind life, the day job and some other projects whose bugs affect me daily. Please feel free to pester me. Since people are apparently using this code, I suppose that I should get busy starting in Q2 to fix this.

Also: can you add some details on the size of your store, actions that cause the fault. I have about 5TB and backups are good, individual restores are good and large restores are too slow. (I just shard them with the shallow feature and it's been fine.)

Aside: I attempted the `rclone` storj backend before writing this patch and it was a (literal) non-starter.

If you have some good tests that show failures, consider adding to the PR?

Aside: @mathrick's comment suggests that `kopia` is hitting storj too hard. I would have expected storj to handle that internally.

**Comment by mathrick** [comment]
@rjkroege: yes, the [limits section](https://storj.dev/dcs/api/s3/s3-compatibility#limits) in the docs definitely doesn't convey that it will just error out, rather than returning results up to the limit with some sort of pagination functionality. This unfortunately completely cooks my backups.

> I have about 5TB and backups are good, individual restores are good and large restores are too slow. (I just shard them with the shallow feature and it's been fine.)

Would you mind elaborating? How do you shard things and what is the shallow feature? Is that a feature of Storj or `kopia`?

**Comment by ptandler** [comment]
> I intend to finish this but it's unfortunately behind life, the day job and some other projects whose bugs affect me daily. Please feel free to pester me. Since people are apparently using this code, I suppose that I should get busy starting in Q2 to fix this.

Yeah, I can fully understand ...

> Also: can you add some details on the size of your store, actions that cause the fault. I have about 5TB and backups are good, individual restores are good and large restores are too slow. (I just shard them with the shallow feature and it's been fine.)

In my storj bucket are >= 20k files and > 440 GB, not that much

> Aside: I attempted the `rclone` storj backend before writing this patch and it was a (literal) non-starter.

I now also tried to connect to my repo via rclone. I finally got rclone config right so I can list the repo files via rclone, but I couldn't yet convince kopia to connect to this repo. Seems, it's looking for a `kopia.repository.f`, but in the file list there's only `kopia.repository`.

So when I wanna try your backend, I pull you PR's branch and try building kopia on my own, I guess? Never work with Go so far, let's see.

**Comment by rjkroege** [comment]
Re "sharding" from the `kopia` help output:

```
If the '--shallow' option is provided, files and directories this depth and below in the directory hierarchy will be represented by compact placeholder files of the
form 'entry.kopia-entry' instead of being restored. (I.e. setting '--shallow' to 0 will only shallow restore.) Snapshots created of directory contents represented by
placeholder files will be identical to snapshots of the equivalent fully expanded tree.

In the expanding-a-placeholder mode:

The source to be restored is a pre-existing placeholder entry of the form 'entry.kopia-entry'. The target will be 'entry'. '--shallow' controls the depth of the
expansion and defaults to 0. For example:

'restore d3.kopiadir'

will remove the d3.kopiadir placeholder and restore the referenced repository contents into path d3 where the contents of the newly created path d3 will themselves
be placeholder files.
```

So what I do for a large restore (I have one). E.g. for this snapshot

```
 2026-03-05 13:01:51 EST kBLAH 3.2 TB drwxr-xr-x files:533546 dirs:231929 (latest-1,hourly-1,daily-1,weekly-1,monthly-1,annual-1)
```

I do

```
kopia restore --shallow=0 kBLAH /restore/here
```

(Maybe it should be `--shallow=1`, I don't remember my own code.) It restores fast. I can then kick off restores of the placeholder files of what's needed. It's been working with my patch: Storj works as long as I don't do too much at once. So I can write a shell script to grab the parts that I most want with a sequence of restores.

**Comment by mathrick** [comment]
@rjkroege: thank you. So this has no bearing on the backups, only restores? Unfortunately I never get to the point I could even contemplate a restore, since `kopia` fails to connect to the repo in the first place. I'm guessing it's your patch that fixes that then, rather than a preexisting option.
